sound/soc/sof/core.c

```diff
@@ -340,9 +340,17 @@ EXPORT_SYMBOL(snd_sof_device_probe);
 
 int snd_sof_device_remove(struct device *dev)
 {
+	int ret;
 	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
 	struct snd_sof_pdata *pdata = sdev->pdata;
 
+	ret = snd_sof_dsp_pre_remove(sdev);
+	if (ret < 0) {
+		dev_err(dev, "error: %d failed to prepare DSP for device removal",
+			ret);
+		return ret;
+	}
+
```

Comment:
@mrajwa Why not do this as part of remove sequence, e.g. in hda.c:hda_dsp_remove()? Not sure why the extra new method is needed and are the expected semantics. If it's called always before remove, what's the benefit? If some delay is needed between sram gating and power-off, this can be added in hda_dsp_remove() as well?

Reply:
I also don't get why we need a new callback. If this is part of the power-down sequence, then let's add it as part of the power down sequence. this should apply to ALL D3 transitions.

Reply:
+1 but @mrajwa this is meant to specifically fix the module reload problem on GLK isnt it? But we've still got the failure in the CI test?

Reply:
@kv2019i, at hda_dsp_remove() time you have IPCs already disabled + machine driver is removed - you won't be able to send IPC message to DSP anymore. Also, I think notifying DSP should be the very first thing to do.
@plbossart, it is power-down related but doesn't have anything to do with power optimization. This PR is the response to this bug  https://github.com/thesofproject/sof/issues/2646. Currently, in kernel you remove the driver without notifying DSP about it. This is wrong per se, plus there are HW limitations that DSP have to take care of. One of them happens when you reset controller without power gating memory - this is the essence of the bug I quoted above.
@ranj063, yes this is the fix. CI will still be failing as there is corresponding DSP patch https://github.com/thesofproject/sof/pull/2837

Reply:
@marjwa then we do need an ABI change if we're adding a new IPC?

Reply:
@plbossart @ranj063 @mrajwa Power down sequence in FW already takes care of disabling LPSRAM. I don't know what is the flow in the kernel, but if you would put DSP into D3 before unloading the module, then this problem should be resolved automagically.

Reply:
@tlauda We dont transition the DSP to D3 the same way we do during a suspend when we remove the module. From a driver point of view, in fact when we remove the module, the SOF core device would in fact remain "active" from a runtime PM point of view. 

But, we do power down the cores and the HDA controller when we remove the module. Do you think it would help if we sent the CTX_SAVE IPC before removing the module maybe?

Reply:
@ranj063, yes it would work also. If there are no other constraints which forbid us from moving DSP to D3 during module removal then I can rework it this way.

Reply:
@mrajwa On a coarse-grained level, the only missing thing I see between the 2 sequences for suspend and remove, is the missing CTX_SAVE IPC. I can do a quick check with it today and update the bug if you like

Reply:
I will finish fixing the bug since I already got involved. However we have to agree if on moving DSP into D3 during module remove is a valid option or we want to do it in "power-state free" maner which is my current patch.

Reply:
Sounds good. I think it is the right thing to do ie put the DSP in D3. Pulling the plug without notifying the DSP doesnt sound right

Reply:
In general the remove should do the same sequence as shutdown, plus free all allocated resources. Thanks for spotting this @mrajwa

Reply:
Better to make it generic and benefit for BYT/CHT/BSW/BDW also. When spotted BSW unload/reload hang issue, we found this is necessary.

Reply:
I have further updated this patch. Now we make use of CTX_SAVE IPC this means no ABI change is needed.

Reply:
@mrajwa better to call it something like notify_dsp() or something else along those lines. Calling it pre_remoce() and calling it inside remove() is not intuitive

Reply:
I don't think we should have support for any dedicated callback specifically for Linux driver removal. This should be part of the standard power down / shutdown procedure, not something extra we only do during driver removal.

Reply:
@plbossart I agree that there's no need for any callbacks here for pre-remove. But there is no standard power-down sequence in the driver. I think any time all cores are powered down, the DSP should be notified in advance. We already do this for suspend and simply adding this to the remove() callback would suffice.

Reply:
@mrajwa I don't think your solution would work. There is a big fat comment in the code that reads:

````c
	/*
	 * Unregistering the machine driver results in unloading the topology.
	 * Some widgets, ex: scheduler, attempt to power down the core they are
	 * scheduled on, when they are unloaded. Therefore, the DSP must be
	 * removed only after the topology has been unloaded.
	 */
	if (sdev->fw_state > SOF_FW_BOOT_NOT_STARTED)
		snd_sof_remove(sdev);
````

Therefore the only 'safe' place to send this CTX_SAVE is likely from snd_sof_remove()

@ranj063 do you concur?

Reply:
@plbossart yes, it is not safe to notify that the DSP is being powered down until it is just before it is actually done. We send IPCs to stop trace when stopping it and those will all fail miserably.

Reply:
> Therefore the only 'safe' place to send this CTX_SAVE is likely from snd_sof_remove()

likely not. The comment says that in `snd_sof_machine_unregister()` which **proceeds** `snd_sof_remove(sdev);` some widgets may power down the DSP core on its own... and you want to send IPC message to DSP in `snd_sof_remove` when there is a change that DSP has already been power gated?
Not to mention before `snd_sof_remove(sdev);` you disable IPCs.... just a few lines above.

> @mrajwa I don't think your solution would work.
Before I push a solution I test it first. Also it has been tested by @sinahuang and proved to work here https://github.com/thesofproject/sof/issues/2646

> We send IPCs to stop trace when

@ranj063  where do you see IPC message to disable trances there? Or rather, do you see any IPC message to DSP in whole .remove procedure prior to my commits? 

I don't see any danger in this solution -> we notify DSP so it can shut itself down gracefully, then we unload topology and other stuff on driver side (no interaction with DSP) and finally we power down various DSP resources but this time in right order.

Reply:
@mrajwa snd_sof_free_trace() will first send a STOP IPC before freeing the dma pages for trace.

Reply:
If you want to keep it here, then the best place is to do this before snd_sof_remove() and after snd_sof_machine_unregister().

Reply:
@mrajwa I also think we cannot do this in the core for the legacy platforms. Those platforms do no support runtime PM and we've never tested the CTX_SAVE IPC with them. So we need to be careful here. Take a look at the CI results

Reply:
@mrajwa any updates? This is a CI blocker

Reply:
@ranj063, as for 

> I think we should do this only for GLK+ in the remove op

This was the initial version, but I took Keyon's concern into the account. He suggested that we should informa all platforms. However I think, since they don't define .remove procedure there is no need to inform them right?